packages/ts/src/components/donut/index.ts

```diff
@@ -57,38 +57,43 @@ export class Donut<Datum> extends ComponentCore<Datum[], DonutConfig<Datum>, Don
 
   _render (customDuration?: number): void {
     const { config, datamodel, bleed } = this
-    const data = datamodel.data
-    const duration = isNumber(customDuration) ? customDuration : config.duration
 
+    // Wrap data to preserve original indices
+    const data: DonutInputArc<Datum>[] = datamodel.data.reduce((acc, d, i) =>
```

Comment:
@reb-dev I find this slightly more difficult to read comparing to the regular `map`  and `filter` approach.

```ts
const data: DonutDatum<Datum>[] = datamodel.data
  .map((d, i) => ({
    ...d,
    _index: i
  })
  .filter(d => !config.showEmptySegments && !getNumber(d, config.value, d._index))
```

P.S. See my comments regarding `DonutDatum` below

Reply:
@rokotyan I agree it's easier to read, but reduce is superior performance-wise so I usually prefer it. Your call though

Reply:
@reb-dev Usually it is, but I don't think it's the case here:

```ts
[...acc, { _datum: d, _index: i }]
```
We're initializing a new array on every step, while memory allocation is an expensive operation which often becomes a performance bottleneck.